### ISSUE TITLE
Potential Regression in Coverage Checker

### ISSUE THREAD

**Issue by jfdm** [idris-lang/Idris2#1891]
Consider the code [in this gist](https://gist.github.com/jfdm/713eb26592b05a074d4de63f87f07ed7).

This code used to type-check, presenting a simple proof-of-progress on projection on tuples.
Now, however, the coverage checker tells me that the code is not covering.

At somepoint in time Idris has changed its mind. So who is correct here: me or Idris?

# Steps to Reproduce

Compile it.

# Expected Behavior

The in-progress proof-of-progress should pass the coverage checker for the hole to be filled in.
The old equation is presented.

# Observed Behavior

The in-progress proof-of-progress presented complains about missing patterns, AFAIK there are no missing patterns.
Adding in a place-holder pattern satisfies the coverage checker, allows the proof to be given albeit with the same hole that needs filling.

**Comment by ohad** [comment]
Thanks!

Can you see a way to minimise it further and still get the error?

**Comment by jfdm** [comment]
I am trying to too...

**Comment by gallais** [comment]
Can't you have a `Proj` too in that case split?

**Comment by jfdm** [comment]
RIght, I am not saying that the term should be be impossible to construct but I am reducing the terms and when expanding proof that tuple is a value then it should be fully reduced.

**Comment by gallais** [comment]
Then don't you need to enforce that in the type, introducing a notion of normal form?

**Comment by jfdm** [comment]
Yes, you had me confused there for a second. I have defined the value form, the issue is that when case splitting on `Done tuple` it *should* rule out the other reducible patterns, but it is not.

**Comment by jfdm** [comment]
So the problem is that when I last tried this, Idris was a bit more loose in what it would accept. It is now a bit more strict, and the computation in the type of `Proj` (aka the 'green slime') can now confuse the checker. 

The problem can be addressed by moving (fording) the computation from being 'within the type' to being 'along side the type'.
We do this by using a secondary datatype that synthesizes the required type.

So given an 'index' type that captures the indexing computation:

```idris
public export
data Index : (idx : Fin n) -> (Vect n a) -> a -> Type where
    Here  : Index    FZ  (x::xs) x

    There : Index     n      xs  x
         -> Index (FS n) (y::xs) x
````
we can redefine the `Proj` constructor as follows.

```idris
Proj : {types : Vect (S n) Ty}
    -> (tuple : Tupled (TyTuple types))
    -> (idx   : Index i types ty)
             -> Tupled ty
```

Modulo updating the remainder of the gist, everything will now be seen as total.

**Comment by gallais** [comment]
> We do this by using a secondary datatype that synthesizes the required type-level value.

Fording is stricto sensu only the conversion of `F (f xs)` to `F a` where `a` is constrained
to be equal to `f xs`. Using a relational definition of `index` instead of a functional one
is another kind of move that makes proving properties by induction over the `index` easier.
